### PR TITLE
[TF-TRT] Add LogSoftmax Support for TF-TRT

### PR DESCRIPTION
**Pull request by pavanimajety**
This PR adds TF-TRT support to `tf.nn.log_softmax` operation.  This is performed using the formula `logsoftmax = logits - log(reduce_sum(exp(logits), axis=-1))` .  The implemented TRT layers are fused into a single op.   

@DEKHTIARJonathan @tfeher : Please review the changes.

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 1]
```diff
@@ -4721,6 +4721,58 @@ Status ConvertSoftmax(OpConverterParams* params) {
...
+  const int num_trt_dims = logits_tensor->getDimensions().nbDims;
+  if (!num_trt_dims && params->use_implicit_batch) {
+    return errors::InvalidArgument(
+        "TensorRT LogSoftmax cannot apply on the batch dimension");
+  }
+
+  if (params->validation_only) return Status::OK();
+
+  // Perform LogSoftmax operation -
+  // logsoftmax = logits - log(reduce_sum(exp(logits), axis))
```
Comment (tfeher) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:4742]: Is there a reason to prefer this formulation over  `log(softmax(logits))`?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 2]
```diff
@@ -4721,6 +4721,58 @@ Status ConvertSoftmax(OpConverterParams* params) {
   return Status::OK();
 }
 
+Status ConvertLogSoftmax(OpConverterParams* params) {
```
Comment (tfeher) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:4724]: Instead of adding new converters to convert_nodes.cc, please prefer to:
- derive it from [OpConverterBase](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2tensorrt/convert/op_converter.h)
- place it under [convert/ops](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2tensorrt/convert/ops)
- Place the validation code into `Validate()`, and the actual conversion into `Convert()`, see the [tile converter](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc) as an example
- don't forget to add an entry to [BUILD](https://github.com/tensorflow/tensorflow/blob/cf29afd3e937e65cdc8d4dff824bcc0af09ab43d/tensorflow/compiler/tf2tensorrt/BUILD#L682)

**Review by tfeher — CHANGES_REQUESTED** [review]
Hi Pavani, thank you for the PR, it looks great overall! But for new converters we prefer to derive them from `OpConverterBase`, I have left instructions on how to update.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc (hunk 1 above)
Comment (pavanimajety) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:4742]: I chose this way because tensorflow implements in Python using `logsoftmax = logits - log(reduce_sum(exp(logits), axis))` .

**Review by tfeher — APPROVED** [review]
Thanks @pavanimajety for refactoring the converter, the PR looks good to me!

**Comment by bixia1** [comment]
@pavanimajety Please add a PR description.

**Review by bixia1 — CHANGES_REQUESTED** [review]
Looks good to me, just request fix to comment style and a PR description.

File: tensorflow/compiler/tf2tensorrt/convert/ops/log_softmax.cc [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+  Status Convert() {
+    const auto &params = *this->params_;
+    const auto &inputs = params.inputs;
+    const auto &node_def = params.node_def;
+
+    // Perform LogSoftmax operation -
+    // logsoftmax = logits - log(reduce_sum(exp(logits), axis))
+
+    // get the logits tensor
```
[the remark this answers is not in the document]
Comment (pavanimajety) [reply tensorflow/compiler/tf2tensorrt/convert/ops/log_softmax.cc]: Fixed the comments.

**Review by bixia1 — COMMENTED** [review]
Looks good to me now. Would you please squash the commits?

**Comment by pavanimajety** [comment]
> Looks good to me now. Would you please squash the commits?

Done. Thanks for reviewing

**Review by bixia1 — APPROVED** [review]

**Comment by pavanimajety** [comment]
Sorry, rebased for further work by mistake.

**Comment by bixia1** [comment]
This is merged, wonder why it doesn't show as merged here.

**Comment by pavanimajety** [comment]
> This is merged, wonder why it doesn't show as merged here.

Would you mind checking again? I don't see in tensorflow:master : https://github.com/tensorflow/tensorflow/tree/master/tensorflow/compiler/tf2tensorrt/convert/ops

**Comment by mihaimaruseac** [comment]
It did not show as merged here because copybara sync was broken. Fixed in big squash commit 49f97f135a2e1d5d22e60d2a80ec668d53f9708a
